crates/compute/src/cpu/layer.rs

```diff
@@ -424,6 +424,53 @@ impl<F: TowerTop> ComputeLayerExecutor<F> for CpuLayerExecutor<F> {
 
 		Ok(())
 	}
+
+	fn pairwise_product_reduce(
+		&mut self,
+		input: <Self::DevMem as ComputeMemory<F>>::FSlice<'_>,
+		round_outputs: &mut [<Self::DevMem as ComputeMemory<F>>::FSliceMut<'_>],
+	) -> Result<(), Error> {
+		if input.len().count_ones() != 1 {
```

Comment:
[nit] There is a function [is_power_of_two](https://doc.rust-lang.org/std/primitive.usize.html#method.is_power_of_two).

Or maybe better use `strict_log_2`, e.g.:
```
match strict_log_2(input.len()) {
    None => Err(Error::InputValidation(format!(
				"input length must be a power of 2: {}",
				input.len()
			))),
   Some(0) => Err(Error::InputValidation(format!(
				"input length must be greater than or equal to 2 in order to perform at least one reduction: {}",
				input.len()
			))),
   Some(log_num_inputs) => {...}
}
```